### DISCUSSION TITLE
GPU memory cannot be released after MXPredFree

### DISCUSSION THREAD

**Discussion by rrjia** [apache/incubator-mxnet#19862] [Q&A] [answered]
Hi,

I use the C++ APIs of MXNet to do some classification tasks. But I find that MXNet does not release GPU memory when calling MXPredFree(). When my program exits, all occupied GPU memory is released. So when I run my program, I can do classification just once, otherwise GPU memory is not enough. I think this is a very serious problem. I hope MXNet developers @tqchen @mli @piiswrong can look into this issue and resolve it ASAP. Thanks!

The sequence of function calls is as follows:
(1) MXPredCreate
(2) MXPredSetInput
(3) MXPredForward
(4) MXPredGetOutput
(5) MXPredFree

after i call MXPredCreate the GPU memory will capture by mxnet, by i can not release gpu memory bu mxnet c_predict_api

**Comment by szha** [comment, marked as answer]
You can use the C-API `MXStorageEmptyCache(dev_type, dev_id)` to ask the memory pool to release the memory

**Reply by rrjia** [reply]
i use MXStorageEmptyCache(dev_type, dev_id) after MXPredFree, the GPU memory from 849M reduce to 703M, but it can not release all GPU memory,  thank you 
```
int ret = MXPredFree(pred_hnd_);	
if (0 != ret)
{
    return -1;
}
ret = MXStorageEmptyCache(device_type_, device_id_);
pred_hnd_ = nullptr;
```
